File: 백준/Gold/9084. 동전/동전.java
import java.io.*;
import java.util.*;
// m 원을 만들 수 있는 방법 수
/**
 * m 원을 만드는 케이스 
 * m - k 원에서 k원을 추가?
 * d[m] += d[m - k]; -> 순서로 인한 경우의 수가 뻥튀기됨
 * 구성의 관점에서 바라봐야 할 듯
 * 
 * 중요: 동전에 순서를 부여해서 과거의 경우의 수를 이용할 때, 순서로 인한 뻥튀기를 막는다.
 * i번째 동전을 사용하지 않고 금액을 구성한 경우 vs i번째 동전을 사용하고 구성한 경우로 나눈다.
 * 
 */

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();

        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());
            int[] costs = new int[n + 1];
            
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 1; i <= n; i++) {
                costs[i] = Integer.parseInt(st.nextToken());
            }
            int m = Integer.parseInt(br.readLine());
            int[][] dp = new int[n + 1][m + 1];
            for (int i = 1; i <= n; i++) {
                dp[i][0] = 1;
                for (int j = 1; j <= m; j++) {
                    dp[i][j] = dp[i - 1][j];
                    if (j - costs[i] < 0) continue;
                    dp[i][j] += dp[i][j - costs[i]];
                }
            }
            sb.append(dp[n][m]).append('\n');
        }
        System.out.print(sb);
    }
    
}

File: 백준/Gold/9084. 동전/동전.kt
/**
 * 금액을 구성하는 모든 방법의 수: (금액, 방법)
 * - 금액은 10000원 이하, 동전의 수는 20개 이하
 * - 금액 x원: d[x] += d[x-coin[k]]
 */

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    repeat(readLine().toInt()) {
        val n = readLine().toInt()
        val coins = StringTokenizer(readLine()).let { st -> IntArray(n) { st.nextToken().toInt() } }
        val fin = readLine().toInt()
        val d = IntArray(fin + 1)

        // 이렇게 되면 순서가 영향을 미친다
        // for (money in 1..fin) {
        //     for (k in 0 until n) {
        //         if (money - coins[k] >= 0) d[money] += d[money - coins[k]]
        //     }
        // }

        // 사용할 수 있는 코인을 순차적으로 적용하면서, 구성에 집중한다.
        for (k in 0 until n) {
            if (coins[k] <= fin) d[coins[k]] += 1
            for (money in coins[k]+1..fin) {
                d[money] += d[money - coins[k]]
            }
        }
        println(d[fin])
    }
}